src/skagent/model_analyzer.py

```diff
@@ -0,0 +1,274 @@
+                    _, params = sd
+                    if isinstance(params, dict):
+                        for v in params.values():
+                            if isinstance(v, str):
+                                self._raw_deps[var].append(v)
+                                self._param_deps[v].add(var)
+
+            # dynamics
+            for var, rule in blk.dynamics.items():
+                deps = (
+                    list(rule.iset)
+                    if isinstance(rule, Control)
+                    else self._rhs_symbols(rule)
+                )
+                for d in deps:
+                    self._raw_deps[var].append(d)
+                    if d in self.calibration:
+                        self._param_deps[d].add(var)
+
+            # reward
+            for var, rd in blk.reward.items():
+                deps = self._rhs_symbols(rd) if callable(rd) else []
+                for d in deps:
+                    self._raw_deps[var].append(d)
+                    if d in self.calibration:
+                        self._param_deps[d].add(var)
+
+        # dedupe
+        for tgt in list(self._raw_deps):
+            self._raw_deps[tgt] = sorted(set(self._raw_deps[tgt]))
+
+    def _identify_time_dependencies(self):
+        """Mark (tgt, src) pairs for lag edges via insertion-order heuristic."""
+        for blk in self._blocks:
+            order = list(blk.dynamics.keys())
+            seen = set()
+            for tgt in order:
+                deps = [d for d in self._raw_deps.get(tgt, []) if d in order]
+                for src in deps:
+                    if src == tgt or src not in seen:
+                        self._prev_deps.add((tgt, src))
+                seen.add(tgt)
+
+    def _assemble_edges(self):
+        """
+        Convert raw_deps + prev_deps + param_deps into four classified edge lists.
+        Preserve self-lag edges (p→p) for visualization splitting.
+        """
+        for tgt, deps in self._raw_deps.items():
+            for src in deps:
+                # 1) lag edges, including self-lag
+                if (tgt, src) in self._prev_deps:
+                    self.edges["lag"].append((src, tgt))
+                    continue
+                # 2) drop any remaining self-loop
+                if src == tgt:
+                    continue
+                # 3) param edges
+                if src in self._param_deps:
+                    self.edges["param"].append((src, tgt))
+                # 4) shock edges
+                elif src in self.node_meta and self.node_meta[src]["kind"] == "shock":
+                    self.edges["shock"].append((src, tgt))
+                # 5) instant edges
+                else:
+                    self.edges["instant"].append((src, tgt))
+
+        # dedupe & sort
+        for et in self.edges:
+            self.edges[et] = sorted(set(self.edges[et]))
+
+    def _collect_formulas(self):
+        """Generate human-readable formulas for each variable."""
```

Comment:
This is cool functionality.
I might wind up moving it to a different part of the library, because it's quite general.
No reason not to have it here for now.